File: solutions/csharp/robot-simulator/1/RobotSimulator.cs
using System;
using System.Collections.Generic;

public enum Direction
{
    North,
    East,
    South,
    West
}

public class RobotSimulator
{
    private Direction _direction;
    private int _x;
    private int _y;

    private Dictionary<char, Action> _movements;
    private Dictionary<Direction, Action> _advance;
    private Dictionary<Direction, Direction> _turnRight = 
        new Dictionary<Direction, Direction>()
    {
        {Direction.North, Direction.East},
        {Direction.East, Direction.South},
        {Direction.South, Direction.West},
        {Direction.West, Direction.North}
    };
    private Dictionary<Direction, Direction> _turnLeft = 
        new Dictionary<Direction, Direction>()
    {
        {Direction.North, Direction.West},
        {Direction.East, Direction.North},
        {Direction.South, Direction.East},
        {Direction.West, Direction.South}
    };

    public RobotSimulator(Direction direction, int x, int y)
    {
        _direction = direction;
        _x = x;
        _y = y;

        _advance = new Dictionary<Direction, Action>()
        {
            {Direction.North, () => _y += 1},
            {Direction.East, () => _x += 1},
            {Direction.South, () => _y -= 1},
            {Direction.West, () => _x -= 1}
        };

        _movements = new Dictionary<char, Action>()
        {
            {'A', Advance},
            {'R', TurnRight},
            {'L', TurnLeft}
        };
    }

    public Direction Direction
    {
        get
        {
            return _direction;
        }
    }

    public int X
    {
        get
        {
            return _x;
        }
    }

    public int Y
    {
        get
        {
            return _y;
        }
    }

    public void Move(string instructions)
    {
        foreach (var c in instructions)
        {
            if (_movements.ContainsKey(c))
            {
                _movements[c]();
            }
        }
    }

    public void Advance()
    {
        _advance[_direction]();
    }

    private void TurnRight()
    {
        _direction = _turnRight[_direction];
    }

    private void TurnLeft()
    {
        _direction = _turnLeft[_direction];
    }
}

File: solutions/typescript/robot-simulator/1/robot-simulator.ts
export class InvalidInputError extends Error {
  constructor(message: string) {
    super()
    this.message = message || 'Invalid Input'
  }
}

type Direction = 'north' | 'east' | 'south' | 'west'
type Coordinates = [number, number]

export class Robot {
  #bearing: Direction = 'north'
  #coordinates: Coordinates = [0,0]
  
  get bearing(): Direction {
    return this.#bearing
  }

  get coordinates(): Coordinates {
    return this.#coordinates
  }

  place({x, y, direction}: { x: number; y: number; direction: Direction }) {
    if (!['north', 'south', 'east', 'west'].includes(direction)) {
      throw new InvalidInputError('direction parameter is not a Direction')
    }
    this.#bearing = direction
    this.#coordinates = [x, y]
  }

  evaluate(instructions: string) {
    for(const instruction of instructions) {
      switch(instruction) {
        case 'A': this.#advance(); break;
        case 'R': this.#turnRight(); break;
        case 'L': this.#turnLeft(); break;
      }
    }
  }

  #advance() {
    switch (this.#bearing) {
      case 'north': this.#coordinates[1] += 1; break;
      case 'east': this.#coordinates[0] += 1; break;
      case 'south': this.#coordinates[1] -= 1; break;
      case 'west': this.#coordinates[0] -= 1; break;
    }
  }

  #turnRight() {
    switch(this.#bearing) {
      case 'north': this.#bearing = 'east'; break;
      case 'east': this.#bearing = 'south'; break;
      case 'south': this.#bearing = 'west'; break;
      case 'west': this.#bearing = 'north'; break;        
    }
  }

  #turnLeft() {
    switch(this.#bearing) {
      case 'north': this.#bearing = 'west'; break;
      case 'east': this.#bearing = 'north'; break;
      case 'south': this.#bearing = 'east'; break;
      case 'west': this.#bearing = 'south'; break;        
    }    
  }
}
